lib/Mapper.php

```diff
@@ -863,6 +885,156 @@ public function upsert(array $data, array $where)
+                            //Update the foreign key to match the main entity primary key
+                            $related->set($relation->foreignKey(), $entity->primaryKey());
+                            $lastResult = $relatedMapper->save($related, $options);
+                        }
+                        $relatedIds[] = $related->id;
+                    }
+
+                    foreach ($oldEntities as $oldRelatedEntity) {
+                        if (!in_array($oldRelatedEntity, $relatedIds)) {
+                            $deletedIds[] = $oldRelatedEntity->primaryKey();
+                        }
+                    }
+                }
+                if (count($deletedIds) || $relatedEntities === false) {
+                    $conditions = [$relation->foreignKey() => $entity->primaryKey()];
+                    if (count($deletedIds)) {
+                        $conditions[$relation->localKey().' :in'] = $deletedIds;
+                    }
+                    if ($relatedMapper->entityManager()->fields()[$relation->foreignKey()]['notnull']) {
+                        $relatedMapper->delete($conditions);
+                    } else {
+                        $relatedMapper->queryBuilder()->builder()->update($relatedMapper->table())->set($relation->foreignKey(), null)->where($conditions);
+                    }
+                }
+            } else if ($relation instanceof Relation\HasOne) {
+                $relatedEntity = $entity->relation($relationName);
+                $relatedMapper = $this->getMapper($relation->entityName());
+                if ($relatedEntity === false || $relation->{$relatedEntity->primaryKeyField()} !== $relatedEntity->primaryKey()) {
+                    if ($relatedMapper->entityManager()->fields()[$relation->foreignKey()]['notnull']) {
+                        $relatedMapper->delete([$relation->foreignKey() => $entity->primaryKey()]);
+                    } else {
+                        $relatedMapper->upsert([$relation->foreignKey() => null], [$relation->foreignKey() => $entity->primaryKey()]);
+                    }
+                }
+                if ($relatedEntity instanceof EntityInterface && ($relatedEntity->isNew() || $relatedEntity->isModified())) {
+                    //Update the foreign key to match the main entity primary key
+                    $relatedEntity->set($relation->foreignKey(), $entity->primaryKey());
+                    $lastResult = $relatedMapper->save($relatedEntity, $options);
+                } 
+            }
+        }
+
+        return $lastResult;
+    }
+
+    /**
+     * Save related entities that have been assigned or loaded (only BelongsTo relations)
+     * See saveHasRelations.
+     *
+     * @param \Spot\EntityInterface $entity
+     * @param array $options
+     * @return mixed
+     */
+    public function saveBelongsToRelations(EntityInterface $entity, array $options = [])
```

Comment:
Same as above comment - can we move this to a separate save method on the `Relation\BelongsTo` entity itself?